workflow/src/main/kotlin/com/r3/corda/sdk/token/workflow/flows/UpdateDistributionList.kt

```diff
@@ -0,0 +1,93 @@
+            val tokenPointer: TokenPointer<T>,
+            val oldParty: Party,
+            val newParty: Party
+    ) : FlowLogic<Unit>() {
+        @Suspendable
+        override fun call() {
+            val evolvableToken = tokenPointer.pointer.resolve(serviceHub).state.data
+            val distributionListUpdate = DistributionListUpdate(oldParty, newParty, evolvableToken.linearId)
+            val maintainers = evolvableToken.maintainers
+            val maintainersSessions = maintainers.map { initiateFlow(it) }
+            // Collect signatures from old and new parties
+            val updateBytes = distributionListUpdate.serialize()
+            val ourSig = serviceHub.keyManagementService.sign(updateBytes.bytes, oldParty.owningKey)
+            val signedUpdate = SignedData(updateBytes, ourSig)
+            // TODO This is naive quick fix approach for now, we should use data distribution groups
+            maintainersSessions.forEach {
+                it.send(signedUpdate)
+            }
+        }
+    }
+
+    @InitiatedBy(UpdateDistributionList.Initiator::class)
+    class Responder(val otherSession: FlowSession) : FlowLogic<Unit>() {
+        @Suspendable
+        override fun call() {
+            val distListUpdate = otherSession.receive<SignedData<DistributionListUpdate>>().unwrap {
+                val update = it.verified()
+                // Check that request is signed by the oldParty.
+                check(update.oldParty.owningKey == it.sig.by) {
+                    "Got distribution list update request signed by a wrong party."
+                }
+                // Check that the request comes from that party.
+                check(update.oldParty == otherSession.counterparty) {
+                    "Got distribution list update request from a counterparty: ${otherSession.counterparty} " +
+                            "that isn't a signer of request: ${update.oldParty}."
+                }
+                update
+            }
+            // Check that newParty is well known party.
+            serviceHub.identityService.wellKnownPartyFromAnonymous(distListUpdate.newParty)
+                    ?: throw IllegalArgumentException("Don't know about party: ${distListUpdate.newParty}")
+            if (getDistributionRecord(distListUpdate.linearId, distListUpdate.newParty).isEmpty()) {
+                // Add new party to the dist list for this token.
+                serviceHub.addPartyToDistributionList(distListUpdate.newParty, distListUpdate.linearId)
+            }
+        }
+
+        @Suspendable
+        private fun getDistributionRecord(linearId: UniqueIdentifier, party: Party): List<DistributionRecord> {
+            return serviceHub.withEntityManager {
+                val query: CriteriaQuery<DistributionRecord> = criteriaBuilder.createQuery(DistributionRecord::class.java)
+                query.apply {
+                    val root = from(DistributionRecord::class.java)
```

Comment:
is there a utility function for this? If not maybe we can create one

Reply:
See `getDistributionList` in the workflow module. Maybe we can add another method. Cheers!

Reply:
The method you mentioned returns a list of parties, I want just a single record that matches pair (linearId, party) - it's more optimal than loading a list and searching for party.

Reply:
Ah, I just read first comment, yes, I will add it to utilities.